activestorage/active.py

```diff
@@ -20,36 +22,115 @@ class Active:
     Version 2 will add methods for actual active storage.
 
     """
-    def __init__(self,uri,*args,**kwargs):
-        """ Instantiate in the same way as normal """
+    def __new__(cls, *args, **kwrags):
```

Comment:
Why would we attach these in this way and not as base class methods? Given the implementation depends not only on their presence in Active, but also in the storage implementation, surely we want a controlled vocab of methods external to here, and simply a table of "methods of implementing the controlled vocab" as normal class attributes?

Reply:
A method defining these would be fine, too. The key for me is that the CV needs to be customizable by subclasses, and that means the CV needs to injected into instances either at instantiation time (\_\_new\_\_), or created on demand (method). If there are other ways that allow an external table to available to subclasses I would be very interested to know of them!

For me, the \_\_new\_\_ approach has the advantages of 1) the CV doesn't get re-initialised every time it is accessed; and 2) is more clearly signalled as a CV.

I'm as guilty as the next person when it comes to premature optimization :) but this is library, rather than user, code so that lowers the barrier for me; and it's something I'm used to using in cf-python.

That all said, I don't mind either way!

Reply:
I'm still slightly puzzled ... why do we want to customise the CV by subclasses, given that no method can work if it is not supported in the storage layer ... so you'd never be subclassing unless the storage layer could do it too?

p.s. (`kw*r*ags`?)

Reply:
`kw*r*ags` -> `**kw*r*ags` :laughing: David is using the `super` class so it's cool, Dazed and Confused Matthew McConnaghey-cool :grin:

Reply:
I think that the use case is that when IPSL want to implement their own version of active storage, they might write their own server-side stuff, but want to subclass `Active` to add new methods, or change the method (aggregation) engines, but keep all of the client side goodness.

And I'll tidy up the keyword rags!

Reply:
Ah, ok, that makes sense.  I still don't like it (it's just a bit too much Python magic, and I never like decorating instances not classes), but maybe I'll have to swallow it. So for now, let's do it this way. It can always be undone.